BUG: Merge "cross" will assert on systems with FIPS enforcement

### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the main branch of pandas.


### Reproducible Example

```python
import ssl
import pandas as pd

# if you have a FIPS capable SSL library
ssl.FIPS_mode_set(1)

df = pd.merge(
  pd.DataFrame([...]),
  pd.DataFrame([...]),
  how='cross'
)
```


### Issue Description

When performing a cross merge on a system with [FIPS Mode](https://www.nist.gov/standardsgov/compliance-faqs-federal-information-processing-standards-fips) enabled, it will raise a ValueError when attempting to use a non-secure hashing algorithm (i.e. md5):
```
>       cross_col = f"_cross_{hashlib.md5().hexdigest()}"
E       ValueError: [digital envelope routines] disabled for fips
```
https://github.com/pandas-dev/pandas/blob/main/pandas/core/reshape/merge.py#L1295-L1314

The only workaround is to disable FIPS mode, which is insecure. 

### Expected Behavior

Pandas should be able to perform a cross merge on systems operating in FIPS mode. md5 is a forbidden algorithm in FIPS environments, so the easiest method is to change the md5 hash to a sha256 (or greater)

### Installed Versions

<details>

 python -c 'import pandas as pd; pd.show_versions()'

INSTALLED VERSIONS
------------------
commit           : e8093ba372f9adfe79439d90fe74b0b5b6dea9d6
python           : 3.8.13.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.10.104-linuxkit
Version          : #1 SMP Wed Mar 9 19:05:23 UTC 2022
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : en_US.UTF-8
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.4.3
numpy            : 1.23.1
pytz             : 2022.1
dateutil         : 2.8.2
setuptools       : 63.4.2
pip              : 22.2.2
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : 2.9.3
jinja2           : 3.1.2
IPython          : None
pandas_datareader: None
bs4              : None
bottleneck       : None
brotli           : None
fastparquet      : None
fsspec           : 2022.7.1
gcsfs            : None
markupsafe       : 2.1.1
matplotlib       : 3.5.2
numba            : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : 7.0.0
pyreadstat       : None
pyxlsb           : None
s3fs             : None
scipy            : None
snappy           : None
sqlalchemy       : 1.4.40
tables           : None
tabulate         : 0.8.10
xarray           : None
xlrd             : None
xlwt             : None
zstandard        : None

</details>

--- a/doc/source/whatsnew/v1.5.0.rst
+++ b/doc/source/whatsnew/v1.5.0.rst
@@ -1095,6 +1095,7 @@ Reshaping
 - Bug in :func:`concat` not sorting the column names when ``None`` is included (:issue:`47331`)
 - Bug in :func:`concat` with identical key leads to error when indexing :class:`MultiIndex` (:issue:`46519`)
 - Bug in :func:`pivot_table` raising ``TypeError`` when ``dropna=True`` and aggregation column has extension array dtype (:issue:`47477`)
+- Bug in :func:`merge` raising error for ``how="cross"`` when using ``FIPS`` mode in ssl library (:issue:`48024`)
 - Bug in :meth:`DataFrame.join` with a list when using suffixes to join DataFrames with duplicate column names (:issue:`46396`)
 - Bug in :meth:`DataFrame.pivot_table` with ``sort=False`` results in sorted index (:issue:`17041`)
 - Bug in :meth:`concat` when ``axis=1`` and ``sort=False`` where the resulting Index was a :class:`Int64Index` instead of a :class:`RangeIndex` (:issue:`46675`)

--- a/pandas/core/reshape/merge.py
+++ b/pandas/core/reshape/merge.py
@@ -10,6 +10,7 @@ import hashlib
 import string
 from typing import (
     TYPE_CHECKING,
+    Callable,
     Hashable,
     cast,
 )
@@ -30,6 +31,7 @@ from pandas._typing import (
     Suffixes,
     npt,
 )
+from pandas.compat import PY39
 from pandas.errors import MergeError
 from pandas.util._decorators import (
     Appender,
@@ -1311,7 +1313,16 @@ class _MergeOperation:
             DataFrames with cross_col, the merge operation set to inner and the column
             to join over.
         """
-        cross_col = f"_cross_{hashlib.md5().hexdigest()}"
+        _md5: Callable
+        if PY39:
+
+            def _md5(_hashlib_md5=hashlib.md5):
+                return _hashlib_md5(usedforsecurity=False)
+
+        else:
+            _md5 = hashlib.md5
+
+        cross_col = f"_cross_{_md5().hexdigest()}"
         how = "inner"
         return (
             left.assign(**{cross_col: 1}),

--- a/pandas/core/reshape/merge.py
+++ b/pandas/core/reshape/merge.py
@@ -6,14 +6,13 @@ from __future__ import annotations
 import copy
 import datetime
 from functools import partial
-import hashlib
 import string
 from typing import (
     TYPE_CHECKING,
-    Callable,
     Hashable,
     cast,
 )
+import uuid
 import warnings
 
 import numpy as np
@@ -31,7 +30,6 @@ from pandas._typing import (
     Suffixes,
     npt,
 )
-from pandas.compat import PY39
 from pandas.errors import MergeError
 from pandas.util._decorators import (
     Appender,
@@ -1313,16 +1311,7 @@ class _MergeOperation:
             DataFrames with cross_col, the merge operation set to inner and the column
             to join over.
         """
-        _md5: Callable
-        if PY39:
-
-            def _md5(_hashlib_md5=hashlib.md5):
-                return _hashlib_md5(usedforsecurity=False)
-
-        else:
-            _md5 = hashlib.md5
-
-        cross_col = f"_cross_{_md5().hexdigest()}"
+        cross_col = f"_cross_{uuid.uuid4()}"
         how = "inner"
         return (
             left.assign(**{cross_col: 1}),

